### PR TITLE
[onnx.export] Avoid dict <-> unordered_map implicit copies

### PR DESCRIPTION
**Pull request by gustavla**
This PR is part of an effort to speed up torch.onnx.export (#121422).

- Avoid [implicit copy](https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html#automatic-conversion) between `pybind11::dict` and `std::unordered_map` that
  happens for every node that gets processed. The copy scales with N
  (number of nodes), so this creates a quadratic time complexity.
  Solution is to always use `pybind11::dict`.
- This alone speeds up exports by x2 for large models.
- Resolves (1) in #121422.

(partial fix of #121422)

### CODE REVIEW

**Review by thiagocrepaldi — COMMENTED** [review]
Thank you for the contribution. Have you seen the CI failures related to this change?

```bash
Traceback (most recent call last):
  File "onnx/test_models_onnxruntime.py", line 215, in test_faster_rcnn
    self.run_test(model, ([x1, x2],), rtol=1e-3, atol=1e-5)
  File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 206, in run_test
    _run_test(model, tracing_remained_onnx_input_idx)
  File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 166, in _run_test
    return run_model_test(
  File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 86, in run_model_test
    return verification.verify(*args, options=options, **kwargs)
  File "<@beartype(torch.onnx.verification.verify) at 0x7f16861c1670>", line 352, in verify
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/verification.py", line 875, in verify
    utils._export(
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/utils.py", line 1612, in _export
    graph, params_dict, torch_out = _model_to_graph(
  File "<@beartype(torch.onnx.utils._model_to_graph) at 0x7f16898bbf70>", line 11, in _model_to_graph
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/utils.py", line 1138, in _model_to_graph
    graph = _optimize_graph(
  File "<@beartype(torch.onnx.utils._optimize_graph) at 0x7f16898bb430>", line 86, in _optimize_graph
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/utils.py", line 677, in _optimize_graph
    graph = _C._jit_pass_onnx(graph, operator_export_type)
  File "<@beartype(torch.onnx.utils._run_symbolic_function) at 0x7f16898d[628](https://github.com/pytorch/pytorch/actions/runs/8501842197/job/23286430563?pr=123063#step:18:629)0>", line 93, in _run_symbolic_function
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/utils.py", line 1956, in _run_symbolic_function
    return symbolic_fn(graph_context, *inputs, **attrs)
  File "<@beartype(torch.onnx.symbolic_opset9.prim_if) at 0x7f1689747ee0>", line 36, in prim_if
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/symbolic_opset9.py", line 6858, in prim_if
    torch._C._jit_pass_onnx_block(
RuntimeError: has been invalidated
```

**Comment by gustavla** [comment]
@thiagocrepaldi I'm still looking for a local repro. It passes locally for me (including on a focal/python 3.8/clang10 docker to try to match the test). Is there an easy way to run the test docker locally (the docker image is not publicly accessible)?

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi I'm still looking for a local repro. It passes locally for me (including on a focal/python 3.8/clang10 docker to try to match the test). Is there an easy way to run the test docker locally (the docker image is not publicly accessible)?

@malfet I remmeber you mentioned a way to docker pull CI containers for local repro, but I dont recall the details. Could you help

**Comment by gustavla** [comment]
@thiagocrepaldi I managed to finally repro locally and fix. Please let me know how it looks now.

**Comment by malfet** [comment]
> I remember you mentioned a way to docker pull CI containers for local repro, but I dont recall the details. Could you help

@thiagocrepaldi it's mentioned in ["Use following to pull public copy of the image"](https://github.com/pytorch/pytorch/actions/runs/8501842197/job/23286430563#step:7:9) section:
> docker pull ghcr.io/pytorch/ci-image:pytorch-linux-focal-py3-clang10-onnx-16884143090f148aa56809ceba26eeb699b5dc49

File: torch/csrc/jit/passes/onnx.cpp [hunk 1]
```diff
@@ -15,7 +15,7 @@
 #include <torch/csrc/jit/python/python_ir.h>
 #include <torch/csrc/utils/pybind.h>
 #include <sstream>
-#include <unordered_map>
+#include <unordered_set>
```
Comment (gustavla) [inline torch/csrc/jit/passes/onnx.cpp:18]: unordered_map is not longer used, but this file still uses unordered_set which relied on a transitive include

**Review by justinchuby — APPROVED** [review]

**Pull request closed without merging**